### PR TITLE
feat: add `dtype` option to `LightningModule.load_from_checkpoint`

### LINKED ISSUE #20833: Add a `dtype` option for `load_from_checkpoint`
### Description & Motivation

Hi,
It would be nice to have a `dtype` argument for `load_from_checkpoint` along with the very cool `map_location`. It would allow the user to set the `dtype` automatically via `load_from_checkpoint` without the manual need of doing `.to(dtype)`.

### Pitch

The fix is should be pretty straightforward. Currently we have:
```python
def load_from_checkpoint(
        cls,
        checkpoint_path: Union[_PATH, IO],
        map_location: _MAP_LOCATION_TYPE = None,
        hparams_file: Optional[_PATH] = None,
        strict: Optional[bool] = None,
        **kwargs: Any,
    ) -> Self:
...
```

It would just get one more argument:
```python
def load_from_checkpoint(
        cls,
        checkpoint_path: Union[_PATH, IO],
        map_location: _MAP_LOCATION_TYPE = None,
        type:torch.dtype=None,
        hparams_file: Optional[_PATH] = None,
        strict: Optional[bool] = None,
        **kwargs: Any,
    ) -> Self:
...
```

Given this actually points to `_load_from_checkpoint` https://github.com/Lightning-AI/pytorch-lightning/blob/3d398240d2f62f2ad05e9eff557d2d5cb44f235c/src/lightning/pytorch/core/saving.py#L53

We can add our extra `dtype` argument here too and can easily change this line https://github.com/Lightning-AI/pytorch-lightning/blob/3d398240d2f62f2ad05e9eff557d2d5cb44f235c/src/lightning/pytorch/core/saving.py#L99

to `return model.to(dtype).to(device)`

Hope this is taken into consideration :) 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by gaurav0107**
## What does this PR do?

Adds an optional `dtype` argument to `LightningModule.load_from_checkpoint`
(plumbed through the internal `_load_from_checkpoint`) so that the restored
model's floating-point parameters and buffers can be cast to a target
`torch.dtype` at load time — for example loading an fp32/bf16-trained model
directly as `float16` for faster inference/sampling, without a separate manual
`.to(dtype)` call. This mirrors how the existing `map_location` argument controls
device placement.

The change is strictly opt-in:

- `dtype` defaults to `None`, which preserves the current behavior exactly.
- When provided, the cast is applied via `torch.nn.Module.to(device=..., dtype=...)`,
  which only casts floating-point tensors and leaves integer buffers unchanged.
- It is added for `LightningModule` only; `LightningDataModule` carries no
  tensors to cast.

Downcasting (e.g. `float32` -> `float16`) reduces numerical precision, so the
behavior is documented in the docstring and left entirely to the user's choice.

Fixes #20833

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? Yes — requested in #20833.
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (docstring updated)
- [x] Did you write any **new necessary tests**? (`test_load_from_checkpoint_dtype`, CPU-only)
- [ ] Did you verify new and **existing tests pass** locally with your changes? `ruff check`/`ruff format --check` pass locally; the full pytest suite was not run in my environment (no GPU/torch dev install) and is left to CI.
- [x] Did you list all the **breaking changes** introduced by this pull request? None — additive, default `None`.
- [x] Did you **update the CHANGELOG**?

</details>

## PR review

Anyone in the community is welcome to review the PR.

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by deependujha — CHANGES_REQUESTED** [review]
I'm not very sure if it has any real UX improvement: [check comment](https://github.com/Lightning-AI/pytorch-lightning/issues/20833#issuecomment-4840675325)

**Comment by gaurav0107** [comment]
Thanks for taking a look, and fair points — let me answer honestly rather than defend the diff as-is.

You're right on two counts. Terminology: this is a dtype/precision cast, not quantization. And on the training path it's genuinely redundant — `Trainer(precision=...)` owns dtype/autocast there, and I wouldn't want a second, differently-scoped way to set that.

Where I think a case remains is the narrow no-Trainer inference/sampling path, which is what @arijit-hub originally asked for in #20833: load a checkpoint straight into a target dtype instead of a manual post-load `.to(dtype)`.

Being candid about the strongest honest version of that, and where the current diff falls short:

- As pure convenience you're correct — `dtype=` vs `.to(dtype)` is the same one-liner, and that alone isn't worth a new argument.
- The real motivation is peak memory: for a large model, materializing every param in the checkpoint's fp32 and then casting keeps a full fp32 copy alive at the moment of the cast, whereas casting weights as they're loaded never holds the full-precision model. For big checkpoints that's the difference between fitting and OOMing, and a post-hoc `.to(dtype)` can't give it to you.
- But this diff casts after load (`model.to(device, dtype)`), so as written it does NOT yet deliver that — it's exactly the one-liner you describe. If we keep it, I'd move the cast into the load path so it actually earns its place.

So, back to you: would you be open to it scoped strictly to the no-Trainer inference path, implemented as a load-time cast (real memory win) with a docstring/example spelling out that use-case and pointing training users to `Trainer(precision=...)`? Or if a third dtype entry point still isn't worth the API surface even then, I'm happy to close it — genuinely your call either way.

**Comment by deependujha** [comment]
@gaurav0107 thanks for your patience. Went through the code quickly and it makes sense to have it. Thanks again!

**Review by deependujha — APPROVED** [review]

**Comment by gaurav0107** [comment]
Rebased onto the latest `master` to clear the merge conflict (single feature commit, CHANGELOG entry preserved alongside the new `#21775` entry). 

For the record, the previous red checks were not caused by this change:
- `docs-make (pytorch, doctest)` failed at the *Install pandoc & texlive* step (apt exit 100) before `make doctest` ever ran — a transient infra hiccup. The `dtype` docstring adds no new doctest (it uses a plain `Example::` literal block).
- The lone `pl-cpu (ubuntu-22.04, lightning, 3.12.7, 2.5.1)` failure didn't reproduce on macOS or Windows for the same matrix entry, and the new `test_load_from_checkpoint_dtype` is OS-independent; it looks like unrelated flakiness that the rebase past the recent torch>=2.5 fixes should clear.

`ruff check` and `ruff format --check` pass locally. CI is re-running now.
